ims/einstein/iscsi/tgt.py

```diff
@@ -29,70 +41,68 @@ def start_server(self):
-            raise iscsi_exceptions.StartFailedException()
+        command = 'sudo systemctl start tgtd'
+        self._check_status(
+            self._run_command(command),
+            iscsi_exceptions.StartFailedException,
+            "Failed to start TGTd on one or either hosts")
 
     @log
-    def stop_server(self):
+    def ensure_running(self):
         """
-        Stops the tgtd service
+        Checks if the service is running on both iSCSI servers.
+        If it is not running on either of the machines, send
+        start command to both servers.
 
-        :return: None
+        :return: None.
         """
-        try:
-            shell.call_service_command('stop', 'tgtd', 'Dead')
-        except ShellException:
-            raise iscsi_exceptions.StopFailedException()
+        # Get the status of tgtd from both machines
+        command = "systemctl is-active tgtd"
+        output = self._run_command(command)
 
-    @log
-    def restart_server(self):
+        for status in output.values():
+            if status.exit_code != 0:
+                self.start_server()
+                break
+
+    def __generate_config_file(self, target_name):
         """
-        Restarts the tgtd service
+        Generates the config file in conf.d/
 
+        :param target_name: Target for which config file should be created
         :return: None
         """
+        # Create the config file in /tmp/
+        config_file_name = target_name + ".conf"
+        config_file_path = os.path.join('/tmp/', config_file_name)
         try:
-            shell.call_service_command('restart', 'tgtd', 'Running')
-        except ShellException:
-            raise iscsi_exceptions.RestartFailedException()
-
-    @log
-    def show_status(self):
-        """
-        Returns the status of tgtd
+            config = open(config_file_path, 'w')
+            for line in open(self.tgt_template_file, 'r'):
+                line = line.replace('${target_name}', target_name)
+                line = line.replace('${ceph_user}', self.fs_user)
+                line = line.replace('${ceph_config}', self.fs_config_loc)
+                line = line.replace('${pool}', self.fs_pool)
+                config.write(line)
+            config.close()
+        except (IOError, OSError) as e:
+            raise iscsi_exceptions.TargetCreationFailed(str(e))
 
-        :return: Running or Dead or Error as String
-        """
+        # Copy the config file to the iscsi servers
         try:
-            status = shell.get_service_status('tgtd')
-            if status not in ['Running', 'Dead']:
-                return "Error"
-        except shell_exceptions.CommandFailedException:
-            raise iscsi_exceptions.ShowStatusFailed()
+            cmds = self.client.copy_file(
+                config_file_path, self.TGT_ISCSI_CONFIG + config_file_name)
+            if self.secondary_iscsi is not None:
+                joinall(cmds, raise_error=True)
```

Comment:
Why do we need this `joinall` here?

Reply:
Per the docs here https://github.com/ParallelSSH/parallel-ssh#sftp
If I am copying to both iscsi hosts, I want to wait for both threads to finish